onnxscript/function_libs/torch_aten/ops/core.py

```diff
@@ -269,44 +269,80 @@ def aten_any(self: TensorType) -> TensorType:
     raise NotImplementedError()
 
 
-@torch_op("aten::arange")
+def _range_supported(dtype: int) -> bool:
+    """Returns true if the dtype is supported by the ONNX Range op."""
+    return dtype in {
+        DOUBLE.dtype,
+        FLOAT.dtype,
+        INT16.dtype,
+        INT32.dtype,
+        INT64.dtype,
+    }
+
+
+@torch_op("aten::arange", trace_only=True)
 def aten_arange(end: Union[DOUBLE, FLOAT, INT16, INT32, INT64], dtype: int = -1) -> TensorType:
     # arange(Scalar end, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
 
-    # Cast input to double if dtype is specified, because the input dtype may be e.g. bool
-    # which Range does not support. The output type is ensured because the output
-    # is casted to the specified dtype.
-    if dtype != -1:
-        end = op.Cast(end, to=DOUBLE.dtype)
+    # NOTE: trace_only because both if branches need to be the same type, but we have
+    # a cast in the if branch.
 
-    result = op.Range(0, end, 1)
-    if dtype != -1:
-        result = op.Cast(result, to=dtype)
+    if _range_supported(dtype):
+        end = op.Cast(end, to=dtype)
+        zero = op.Cast(0, to=dtype)
+        one = op.Cast(1, to=dtype)
+        result = op.Range(zero, end, one)
+    else:
+        # Cast input to float if dtype is not supported by Range,
+        # because the input dtype may be e.g. bool
+        # which Range does not support. The output type is ensured because the output
+        # is casted to the specified dtype.
+        end_ = end
+        end = op.Cast(end, to=FLOAT.dtype)
+        zero = op.Constant(value_float=0.0)
+        one = op.Constant(value_float=1.0)
+        result = op.Range(zero, end, one)
+        if dtype != -1:
+            result = op.Cast(result, to=dtype)
+        else:
+            result = op.CastLike(result, end_)
```

Comment:
May be it would be better to handle the `dtype == -1` case separately, since it seems doable without the final cast (eg., from float to some int-type). It is almost the same as the then-branch, except we need `CastLike(_, end_)`.

Reply:
there are types Range doesn't support, commented in code

```
        # Cast input to float if dtype is not supported by Range,
        # because the input dtype may be e.g. bool
        # which Range does not support. The output type is ensured because the output
        # is casted to the specified dtype.
```

Does that sound reasonable?

Reply:
Yes. But in the case where `end` has a supported type and `dtype == -1` (which may be a common case) you will end up with more convoluted code with extra casts. So, I should refine my suggestion to this case.

Reply:
But we don't know if end has a supported type I don't think?

Reply:
Unless we create an overload and dispatch from the exporter

Reply:
I thought about this again. You are right. I will update later.

Reply:
Done.